src/Estimators/MagnetizationDensity.h

```diff
@@ -0,0 +1,220 @@
+ *\mathbf{m}_c = \int d\mathbf{X} \left|{\Psi(\mathbf{X})}\right|^2 \int_{\Omega_c}d\mathbf{r} \sum_i\delta(\mathbf{r}-\hat{\mathbf{r}}_i)\int_0^{2\pi} \frac{ds'_i}{2\pi} \frac{\Psi(\ldots \mathbf{r}_i s'_i \ldots )}{\Psi(\ldots \mathbf{r}_i s_i \ldots)}\langle s_i | \hat{\sigma} | s'_i \rangle
+ *
+ * The way that the above relates to the underlying data structure data_ is as follows.  We grid space up and assign an index
+ * for each of the real space bins (identical to SpinDensityNew).  To account for the fact that magnetization is vectorial, we 
+ * triple the length of this array.  If grid_i is the index of the real space gridpoint i, then the data is layed out like:
+ * [grid_0_x, grid_0_y, grid_0_z, grid_1_x, ..., grid_N_x, grid_N_y, grid_N_z].  This is also the way it is stored in HDF5.  
+ *
+ */
+class MagnetizationDensity : public OperatorEstBase
+{
+public:
+  using Value              = QMCTraits::ValueType;
+  using FullPrecValue      = QMCTraits::FullPrecValueType;
+  using Real               = RealAlias<Value>;
+  using FullPrecReal       = RealAlias<FullPrecValue>;
+  using Grad               = TinyVector<Value, OHMMS_DIM>;
+  using Lattice            = PtclOnLatticeTraits::ParticleLayout;
+  using Position           = QMCTraits::PosType;
+  using Integrator         = MagnetizationDensityInput::Integrator;
+  static constexpr int DIM = QMCTraits::DIM;
+
+  MagnetizationDensity(MagnetizationDensityInput&& min, const Lattice& lattice);
+  MagnetizationDensity(const MagnetizationDensity& magdens, DataLocality dl);
+
+  void startBlock(int steps) override;
+
+  void accumulate(const RefVector<MCPWalker>& walkers,
+                  const RefVector<ParticleSet>& psets,
+                  const RefVector<TrialWaveFunction>& wfns,
+                  RandomGenerator& rng) override;
+
+  void collect(const RefVector<OperatorEstBase>& operator_estimators) override;
+  /**
+  * This returns the total size of data object required for this estimator.
+  * Right now, it's 3*number_of_realspace_gridpoints
+  *
+  * @return Size of data.
+  */
+  size_t getFullDataSize();
+  std::unique_ptr<OperatorEstBase> spawnCrowdClone() const override;
+  void registerOperatorEstimator(hdf_archive& file) override;
+
+  /**
+  * This is a convenience function that handles \int_0^{2\pi} dx f(x).
+  * There are two provided methods for integrating this, so this function picks Simpson's
+  * or Monte Carlo to perform this integration, while keeping awareness of the integration
+  * interval and number of samples.
+  *
+  * @tparam Value type.  Real or complex in double or single precision.  Return type is same as fgrid type.
+  * @param[in] fgrid f(x), the function to integrate.  Assumed to be on a [0-2pi] interval, and if the grid isn't
+  *        random, it's assumed to be uniform.
+  * @return Value of integral.
+  */
+  template<class VAL>
+  VAL integrateMagnetizationDensity(const std::vector<VAL>& fgrid)
```

Comment:
I'm honestly not sure what the preferred style is here, but it looks like in other Estimators like the 1rdm, we use explicit template instantiation in the cpp file. So that would mean moving this and other templated functions to the cpp file. 

Also, why does this need to be templated anyway? isn't the real/complex/single/full precision set by VAL determined from QMCTraits::ValueType? maybe I'm overlooking something, but I think this could just be changed to use Value everywhere

Reply:
Eliminating the templating for this function.